File: Binary Search/Tasks/sqrt.py
"""
Algorithm calculates square root for given number
"""


def binary_search(target, precision=1e-9):
    if number < 0:
        return -1

    left = 0
    right = target
    mid = 0

    while right - left > precision:
        mid = left + (right - left) / 2

        if mid ** 2 < target:
            left = mid
        else:
            right = mid

    return mid


if __name__ == '__main__':
    number = 2
    output = binary_search(number)

    print(f'The nearest found square root of number {number} is {output}.')

File: Binary Search/Tasks/sqrt.cpp
/*
Algorithm calculates square root for given number
*/

#include <iostream>
#include <ctype.h>
#include <iomanip>

double binarySearch(uint32_t number, double precision=1e-9);

int main() {
    uint32_t number = 2;
    double output = binarySearch(2);

    std::cout << "The nearest found square root of number " << number << " is " << std::setprecision(10) << output << std::endl;
    return 0;
}

double binarySearch(uint32_t number, double precision) {
    double left = 0, right = number, mid = 0;

    while (right - left > 1e-9) {
        mid = left + (right - left) / 2;
        
        if (mid * mid < number) {
            left = mid;
        } else {
            right = mid;
        }
    }

    return mid;
}
